vm/actor/src/builtin/market/mod.rs

```diff
@@ -4,7 +4,687 @@
+                METHOD_SEND,
+                &Serialized::default(),
+                &amount_slashed_total,
+            )?;
+        }
+        rt.send(
+            &recipient,
+            METHOD_SEND,
+            &Serialized::default(),
+            &amount_slashed_total,
+        )?;
+        Ok(())
+    }
+
+    /// Deposits the received value into the balance held in escrow.
+    fn add_balance<BS, RT>(rt: &mut RT, provider_or_client: Address) -> Result<(), ActorError>
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        let (nominal, _) = escrow_address(rt, &provider_or_client)?;
+
+        let msg_value = rt.message().value().clone();
+        rt.transaction::<State, Result<(), ActorError>, _>(|st, rt| {
+            st.add_escrow_balance(rt.store(), &nominal, msg_value)
+                .map_err(|e| {
+                    ActorError::new(
+                        ExitCode::ErrIllegalState,
+                        format!("adding to escrow table: {}", e),
+                    )
+                })?;
+
+            // ensure there is an entry in the locked table
+            st.add_locked_balance(rt.store(), &nominal, TokenAmount::zero())
+                .map_err(|e| {
+                    ActorError::new(
+                        ExitCode::ErrIllegalArgument,
+                        format!("adding to locked table: {}", e),
+                    )
+                })?;
+            Ok(())
+        })??;
+
+        Ok(())
+    }
+
+    /// Publish a new set of storage deals (not yet included in a sector).
+    fn publish_storage_deals<BS, RT>(
+        rt: &mut RT,
+        params: PublishStorageDealsParams,
+    ) -> Result<PublishStorageDealsReturn, ActorError>
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        let mut amount_slashed_total = TokenAmount::zero();
+
+        // Deal message must have a From field identical to the provider of all the deals.
+        // This allows us to retain and verify only the client's signature in each deal proposal itself.
+        rt.validate_immediate_caller_type(CALLER_TYPES_SIGNABLE.iter())?;
+        if params.deals.is_empty() {
+            return Err(ActorError::new(
+                ExitCode::ErrIllegalArgument,
+                "Empty deals parameter.".to_owned(),
+            ));
+        }
+
+        // All deals should have the same provider so get worker once
+        let provider_raw = &params.deals[0].proposal.provider;
+        let provider = rt.resolve_address(&provider_raw)?;
+
+        let (_, worker) = request_miner_control_addrs(rt, &provider)?;
+        if worker != rt.message().from().clone() {
+            return Err(ActorError::new(
+                ExitCode::ErrForbidden,
+                format!("Caller is not provider {}", worker),
+            ));
+        }
+
+        // All deals should have the same provider so get worker once
+        let provider_raw = &params.deals[0].clone().proposal.provider;
```

Comment:
I think you can clone at the end instead of cloning the whole deal `&params.deals[0].proposal.provider.clone();`